### ISSUE TITLE
More than one CA broken

### ISSUE THREAD

**Issue by tlhackque** [icing/mod_md#199]
The short form: If you configre more than one CA, any errors corrupt the `mod_md` data store.

Long version:

As part of testing, I setup domains with several CAs - only one of which is LE (v2 staging).
These are URLs with dns names that don't even resolve.
I expected contact attempts to create job activity which would be logged.  This works.

But: the driver assumes only one CA.  So when it gives up, it decides to "reset" the staging area.
This deletes all the challenge data for **all** the CAs - good as well as bad.

So when LE asks for the `acme-tls/1` challenge response, it gets errors like:
````
md_store_fs.c(441): (2)No such file or directory: loading type 2 from /etc/httpd/md-staged/challenges/.../acme-tls-alpn-01.cert.pem
md_store_fs.c(441): loading type 3 from /etc/httpd/md-staged/challenges/.../acme-tls-alpn-01.key.pem
````
And then for non-challenge activity, it has deleted the fallback certs.  And the other CA's real certs and keys!

To fix this, the storage needs to be organized by CA.  E.g. the directory tree needs to be something like &lt;ca&gt;/{challenges,staging,tmp} for sure.  accounts already have the ca name as part of the filename.  domains & archive can probably stay as they are - though what happes when a domain moves from one CA to another needs some thought.  It might be simplest to just put the CA name between `MDStore` and the current tree.  The driver would work each CA in turn.

This isn't urgent, but at some point someone will use one (real) CA for some domains, and a different (real) CA for others...  And configuration allows this.  So I didn't want to lose the issue.

Here is the important part of the config:
````
    MDCertificateAuthority "https://acme-staging-v02.api.letsencrypt.org/directory"
    MDStoreDir "md-staged"
# This file includes the fake domains
Include conf/md-test.conf
    MDCertificateProtocol "ACME"
    MDCertificateAgreement "accepted"

    MDPortMap http:-
    MDCAChallenges tls-alpn-01
        Protocols h2 http/1.1 acme-tls/1
 # This one should work
    <MDomain mydomain.net>
        MDMembers auto
    </MDomain>
````
The fake domains are in `md-test.conf`, which is generated by this script:
````
#!/usr/bin/perl
use warnings;
use strict;
use File::Basename;
my $n="Abc0000";
if(open( my $fh, ">",dirname($0).'/'.'md-test.conf')) {
   print( dirname($0).'/'.'md-test.conf', "\n");
        for(my $i=0; $i<100;$i++){
                printf $fh ("<MDomain %s.example.net>\nMDCertificateAuthority https://noapi%u.mydomain.net/directory\nMDRenewMode manual\nMDMembers manual\nMDPrivateKeys P-256 RSA 2048\n", ++$n,1+($i%20));
                for(my $h=0;$h<$i%20;$h++) {
                        printf $fh ("MDMember %s-%02u.example.org\n",$n,$h);
                }
                print $fh  "</MDomain>\n";
         }
         close $fh;
 }
````
If you include this test file, you'll also see that there are some scaling challenges...

**Comment by icing** [comment]
Hm, I have not fully understood the problem. Challenge data is stored per MDomain name. So, the different MDomains you created should all use (and remove) different directories, no?

**Comment by tlhackque** [comment]
I may not have diagnosed it correctly.  May be necessary to go through the log files.

For now, it's low priority - it hurt, so I stopped.

What I ecountered was that at some point, mod_md gave up and removed everything under the staging tree.  For all CAs.  Even though only the "fake" one was not responding.  mod_md got VERY confused, and would fail trying to talk to LE.  It would delete files, try to get a new account unnecessarily  (because that was also staged & deleted).  What it did was reasonable for the failing CA - but not for the "good one".  I have log files at trace6 - a lot was happening :-)

What I now think is that I had successfuly obtained some certs from LE, that were staged.
The fake CA exhausted the retry count, and "reset" the staging area.  In addition to the fake CA data, this deleted the LE certs that were pending install.  And the "new" account for LE.  The alpn challege certs for the LE domain were torn down because the LE certs issued.  Things go down hill from there.  I may be out of LE quota...

I'm not sure if it's currently possible to have one domain under multiple CAs (e.g. when migrating from one to another).  (e.g. &lt;Mdomain example.net host.example.net&gt; CA LE &lt;/Mdomain&gt; and then &lt;Mdomain host.example.net example.net&gt; CA LnE &lt;/Mdomain&gt; )

In any case, the scenario where someone wants to incrementally move from one CA to another needs to be considered.  I think it should be the case that:
 - It's as easy for the user as replacing the `MDCertificateAuthority` + contact
 - The old certs persist somewhere as long as they are valid (change is dangerous)
 - Having a domain in more than one CA could be useful - suppose one CA dies, having certs from another would allow fail-over to a "backup" CA

As I said, not an immediate priority for me.  But as ACME gets adopted more widely, people will migrate for one reason or another.  Free is good, but perhaps another CA will offer a different service - perhaps EV renewals, or e-mail certs.  Free, or not.  Once that happens, migration and backup will become more important. 

That implies that more than one active CA has to work - not just in the regular cases, but also when a mix of failing and working CAs are active.
